Dcm ngm primer (#53)
CDCgov/cfa-ngm-widget: docs/primer.md

# A Primer on Next Generation Matrix Models

A Next Generation Matrix model is a way to model the expected number of infections generated by a typical infected individual in different groups or categories of the population in consecutive generations. The Next Generation Matrix (here after referred to as the NGM) encodes this information. NGM models are an effective way to model average dynamics in a heterogeneous population during the early growth phase and in the limit of the disease-free equilibrium.

> An NGM model is related to the branching process concept of an offspring distribution generated by an individual. In this context, the NGM represents the expected value of the offspring distribution, or in this case, the distribution of infections caused in a group from a typical infectious individual in another group.

Author: I see. Yes a distribution should be normalized and sum to one. My sentence is just too vague. I was referring to the expected value of the distribution, which I've always thought to mean the same as the first moment, but I suppose that's not as clear. I should probably change this to "expected value of the offspring distribution, or in this case, the expected value of the number of infections".

> An NGM model is related to the branching process concept of an offspring distribution generated by an individual. In this context, the NGM represents the expected value of the offspring distribution, or in this case, the distribution of infections caused in a group from a typical infectious individual in another group.

Reviewer: Maybe like this?

```suggestion
An NGM model is related to the branching process concept of an offspring distribution generated by an individual. In this context, with multiple types of individuals, the NGM represents the expected value of the (conditional) offspring distributions from each group to each group. That is, it provides the average number of infections a typical individual in one group will cause in another.
```

This primer is meant to supplement these works and articulate how the NGM can be used in a transmission model in addition to being an analytical tool.

## Use of NGM as a model
Most commonly, NGMs are used in infectious disease modeling as an analytical tool to estimate the potential for growth of a disease in a population. NGMs are particularly useful for this when a population can be split into a finite number of discrete categories with different epidemiologically relevant traits. In that case, we can define the NGM and use it calculate the basic reproduction number $R_0$, a quantity that can provide insight about the early growth of a disease in a population and interventions that may be effective at controlling its growth. $R_0$ can be computed as the spectral radius of the NGM.

As a result, most modelers familiar with NGMs have experience with using them as an analytical tool rather than as a simulation tool. However, NGMs can also be used to approximately model the ODEs for the subsystem of infected states.

## Interpretation of matrix elements

> Imagine we have an NGM, $\mathbf{R} = [R_{ij}]$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric \mathbf{R}.

Reviewer 2: @dinacmistry I think this is an interesting scope question. I think it's useful to have a document that explains what an NGM is, and shows that it's something that can be either asserted & analyzed OR derived from ODEs. @afmagee42 's questions are valid, but I find it hard to imagine that a CFA project is going to include someone computing NGMs from a complex, multicompartment model.

Another way to say it is: my hypothesis is that: the Dieckmann et al. way of thinking is potentially useful for education/demonstration, but it's not the way CFA is actually going to produce or analyze any of its models. (This is where Dylan Morris would write "<ducks>")

> Imagine we have an NGM, $\mathbf{R} = [R_{ij}]$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric \mathbf{R}.

Reviewer: I feel like there's a benefit of being able to understand it both ways. If one knows how an NGM comes out of a compartmental model, then it should be easier to see the limitations and assumptions that go into it from a perspective closer to that of a lot of epidemiologically-trained folks.

As it is right now I get it as a pure exponential growth branching process model, but if you asked me whether the behavior would be different between assuming new infections are immediately infectious or not, I couldn't tell you.

> Imagine we have an NGM, $\mathbf{R} = [R_{ij}]$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric \mathbf{R}.

Reviewer 2: This is a good point: the "forward" way of thinking says "if I have a fixed NGM, then I'll necessarily get exponential growth" while the inference/analytical way of thinking says "I'll can the NGM that emerges from these equations, at a particular instant in time, and linearize around the DFE, to get an approximation of growth rate, knowing that the NGM will necessarily change in the next instant"

Only say that less confusingly

> Imagine we have an NGM, $\mathbf{R} = [R_{ij}]$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric \mathbf{R}.

Author: > I'm more used to seeing curly braces here (R = {R_ij})? But I can't seem to get GHFMD to render those, so, maybe that's why they're square? Note that GHFMD can't render `mathbf` either.

This is more a style thing that I picked up during my undergrad days. Curly braces, parentheses, and square brackets are all acceptable and I've seen different one in textbooks from different fields. I think square brackets might be more common among physicists (which is what I trained in). Clearly I ended up with mixed notation here as I'm using parentheses for the matrices themselves and square for the abbreviation, which isn't confusing at all.

> Imagine we have an NGM, $\mathbf{R} = [R_{ij}]$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric \mathbf{R}.

Author: > I am realizing now that I am actually kind of confused about the relationship this "the NGM is just R" approach and the "get an NGM from a compartmental model" approach.
> 
> The distinction between the two matrices below of small and large domain makes me think that we can only do this for models where there is only a single state at infection for each group. While that's true of the models that come to mind readily, is it guaranteed to be true of all compartmental models with multiple groups? A quick survey produces the models in [figure 6 here](https://pmc.ncbi.nlm.nih.gov/articles/PMC7436714/) and [figure 2f here](https://pmc.ncbi.nlm.nih.gov/articles/PMC7613485/) which make me think not?

I don't think it's a requirement that there's only a single state of infection for each group. A few clarifying questions:

By group, do we mean things like a risk group, or an age group? Some way to split the population into groups with different epidemiological traits? And by state of infection - do you mean a state that can infect others or do you mean just an infected state? For example, in an SEIR model, E is an infected state, but it's typically not modeled as infectious as well.

If you mean groups and infectious states as I defined above, then yes, you could have multiple infectious states + groups in a model and derive the NGM. For example, you could write a model with an asymptomatic infectious state and symptomatic, or you could even write the chained infectious states model and get the NGM. You could do any combination of this and still write down the NGM and I believe the interpretation of the elements as the reproduction number between groups and states still holds. But it does get tedious to write out that matrix as well as less useful as a simulation model. Potentially it's still useful as an analytical tool for $R_0$?

> Imagine we have an NGM, $\mathbf{R} = [R_{ij}]$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric \mathbf{R}.

Author: Thanks for the feedback! Chatted with Andy about some things here that were less than clear from the primer and the literature. This ended up with a note in the primer about how you can have multiple infectious states but that this example only considers a single one. 

I'm going to say that an example with multiple infectious states and multiple subgroups is out of scope and something better for a cfa-recipes primer on NGMs in more depth.

### Depletion of susceptibles

NGM models describe infectious disease dynamics as a demographic process in the sense that each consecutive generation produces new offspring infections. This can be a good approximation for dynamics early on when the population can be roughly described as fully susceptible. However, unlike ODE models, an NGM model does not account for the fixed size of a population and cannot model the depletion of susceptibles over time.

### Other conditions
The NGM must be non-negative to guarantee that $R_0$ will be a single unique, positive real-valued eigenvalue of $\mathbf{R}$.

> ## Formal definition

Author: After some review, I've realized that I perhaps made this more complex than it needed to be and that's probably causing some confusion. 

You can always get $R_0$ from any of the NGM definitions, but if you're looking for an easy way to get analytical expressions for $R_0$, reducing the system down into $R$ with the auxiliary matrix is helpful. This could mean you go from a matrix $R_L$ with dimension 4 down to $R$ with dimension 2 and that's a hell of a lot easier to derive $R_0$ from.

We also assume that average mixing holds for all individuals between the groups and within, i.e. no individual in either group has different contact rates than others in their group. Individuals in the two risk groups can interact with each other in some way such that an infectious individual would generate some number of new infections in the two groups. More specifically, an average infected individual in group $j$ generates $\beta_{ij}$ infections per unit time in group $i$ in a fully susceptible population.

Unlike the example in Keeling & Rohani, here we model the counts of the population in each state rather than the proportion. We are also modeling the effective rate of transmission between groups as split into two factors: a rate of transmission from group $j$ to group $i$, $\beta_{ij}$ and a rate of interaction based on the number of people in the population available for contact with infectious individuals, i.e. $\frac{S_i}{N}$. This follows from the frequency dependent assumption where effective contact structure that generates transmission is independent of population size (the interested reader can refer to Keeling & Rohani, 2008 pp 17-18 for more details).

> At any given time, there is some fraction of the population that is susceptible in group $i$ and can be infected through interaction with an infected individual in group $j$. Then the average number of infections generated by this individual in group $i$ is $\frac{\beta_{ij}S_i}{N}$ per unit time. Assuming no collision of transmission events, $I_j$ infected individuals produce $\frac{\beta_{ij}S_i I_j}{N}$ infections per unit time.

Author: Yep, I see it now. Clarifying the language to avoid this confusion. Thanks!

Individuals in each risk group also recover from infection at some rate $\gamma_i$. Here we assume that individuals in both groups recover at the same rate, however the following can be generalized to scenarios where average recovery rates of the two groups are different.

Now we can write the infected subsystem of differential equations as

$\frac{d I_H}{dt} = \frac{\beta_{HH}S_H I_H}{N} + \frac{\beta_{HL}S_H I_L}{N} - \gamma I_H$

$\frac{d I_L}{dt} = \frac{\beta_{LH}S_L I_H}{N} + \frac{\beta_{LL}S_L I_L}{N} - \gamma I_L$

or more concisely as

$\frac{d I_i}{dt} = \sum_{j} \frac{\beta_{ij}S_i I_j}{N} - \gamma I_i$

From here we can decompose the system into transmission and transition components, $\mathbf{T}$ and $\mathbf{\Sigma}$, respectively.

Let
$\mathbf{x} =
\left(\begin{array}{cc}
I_H\\
I_L
\end{array}\right)$,

> $\mathbf{T} = [T_{ij}]$ with $T_{ij} = \frac{\beta_{ij}N_i}{N}$, and $\mathbf{\Sigma} = \gamma \mathbb{1}$

Author: Made 3 changes here: 

- correction to sign in Sigma. Good catch!
- using \mathbb{I}_2 instead because that's an even more explicit way to write the identity matrix and hopefully no one confuses it with the infectious state I
- added a line about linearization that turns all $S_i$ into $N_i$

Then we can write the infected subsystem as $\mathbf{\frac{dx}{dt}} = (\mathbf{T} + \mathbf{\Sigma})\mathbf{x}$. The NGM can be defined as

$R = -\mathbf{E}'\mathbf{T}\mathbf{\Sigma}^{-1}\mathbf{E}$.

For this system, the auxiliary matrix is
$\mathbf{E} =
\left(\begin{array}{cc}
1 & 0\\
0 & 1
\end{array}\right

> )$

Author: Out of scope, but duly noted and this will go into cfa-recipes primer!

### Depletion of susceptibles

NGM models describe infectious disease dynamics as a demographic process in the sense that each consecutive generation produces new offspring infections. This can be a good approximation for dynamics early on when the population can be roughly described as fully susceptible. However, unlike ODE models, an NGM model does not account for the fixed size of a population and cannot model the depletion of susceptibles over time.

> ### Other conditions

Author: Scott's PR addresses this. Will leave as is for now.

> The NGM must be non-negative to guarantee that $R_0$ will be a single unique, positive real-valued eigenvalue of $\mathbf{R}$.

Reviewer 3: ```suggestion
Entries of the NGM must be non-negative to guarantee that $R_0$ will be a single unique, positive real-valued eigenvalue of $\mathbf{R}$.
```

Author: I missed this in re-reading the Diekmann article, but they mention other conditions. Not specifically about the NGM when you interpret it directly as R, but conditions on T and Σ that are required for the decomposition. 

```
For completeness we remark that in the decomposition T + Σ it is essential only that T is a non-negative matrix and that Σ is a positive off-diagonal matrix with spectral bound s(Σ)< 0
```

At any given time, there is some fraction of the population that is susceptible in group $i$ and can be infected through interaction with an infected individual in group $j$. Then the average number of infections generated by this individual in group $i$ is $\frac{\beta_{ij}S_i}{N}$ per unit time. Assuming no collision of transmission events, $I_j$ infected individuals produce $\frac{\beta_{ij}S_i I_j}{N}$ infections per unit time.

Individuals in each risk group also recover from infection at some rate $\gamma_i$. Here we assume that individuals in both groups recover at the same rate, however the following can be generalized to scenarios where average recovery rates of the two groups are different.

Now we can write the infected subsystem of differential equations as

$\frac{d I_H}{dt} = \frac{\beta_{HH}S_H I_H}{N} + \frac{\beta_{HL}S_H I_L}{N} - \gamma I_H$

$\frac{d I_L}{dt} = \frac{\beta_{LH}S_L I_H}{N} + \frac{\beta_{LL}S_L I_L}{N} - \gamma I_L$

or more concisely as

$\frac{d I_i}{dt} = \sum_{j} \frac{\beta_{ij}S_i I_j}{N} - \gamma I_i$

From here we can decompose the system into transmission and transition components, $\mathbf{T}$ and $\mathbf{\Sigma}$, respectively.

Let

> $
> \mathbf{x} =
> \left(\begin{array}{cc}
> I_H\\
> I_L
> \end{array}\right)
> $,

Reviewer 4: I don't see this rendering properly, you can (a) use the `math` block or (b) do the following:

```suggestion
$\mathbf{x} = \left(I_H \quad I_L\right)^{\text{t}}$,
```

The same happens with the equations after this one.

Author: That's odd, how are you viewing this? I've been opening with preview in VSCode and I'm not running into any rendering issues. Let me know how so I can reproduce the issue, try your suggestion, and verify that it works. Otherwise, there might be too much back and forth on how to fix this. Thanks!

Author: Nevermind, I see it in github's preview now

Author: I know we can do the math block option but I'm trying to see if there's a way to do the math inline so bear with me for a little while I figure this out. Thanks! Not super familiar with `mkdocs` - if we went that route, would the math render properly when viewed in GitHub? I think that's the important bit here for online readability.

Reviewer: Github can only render github-flavored MD, but `mkdocs` can do more. So it wouldn't render _here_, but it could render on a documentation website.

We do this in [cladecombiner](https://github.com/CDCgov/cladecombiner). With a few lines of configuration in [mkdocs.yml](https://github.com/CDCgov/cladecombiner/blob/26ce0d21a8915c2f16360b584c1b2d92b7b20b10/mkdocs.yml#L3) and a few files in [`.github/actions` and `.github/workflows`](https://github.com/CDCgov/cladecombiner/commit/26ce0d21a8915c2f16360b584c1b2d92b7b20b10), everything in `/docs` gets built into [a website](https://cdcgov.github.io/cladecombiner/) alongside the API reference automatically, (and checked on PRs).

Author: Matrices are rendering properly now but this was a headache and I imagine other math related things might not be so easy to fix later on so I made an issue for us to consider using mkdocs instead.

$\mathbf{T} = [T_{ij}]$ with $T_{ij} = \frac{\beta_{ij}N_i}{N}$, and $\mathbf{\Sigma} = \gamma \mathbb{1}$

Then we can write the infected subsystem as $\mathbf{\frac{dx}{dt}} = (\mathbf{T} + \mathbf{\Sigma})\mathbf{x}$. The NGM can be defined as

$R = -\mathbf{E}'\mathbf{T}\mathbf{\Sigma}^{-1}\mathbf{E}$.

For this system, the auxiliary matrix is
$
\mathbf{E} =
\left(\begin{array}{cc}
1 & 0\\
0 & 1
\end{array}\right
)
$
with unit vector $
\left(\begin{array}{cc}
1\\
0
\end{array}\right)
$ for the state $I_H$ and unit vector $
\left(\begin{array}{cc}
0\\
1
\end{array}\right)
$ for the state $I_L$ in the transmission matrix $\mathbf{T}$.

> Then the NGM can be defined as $\mathbf{R}$ with elements $R_{ij} = \frac{\beta_{ij}N_i}{\gamma N}$. This is the formulation used for the NGM model in this repository.

Reviewer 3: Did we decided to generalize this NGM primer away from the widget? If not, it feels like there is a small disconnect between what they actually enter in the widget (which assumes population sizes have already been factored into the entries of the NGM) and this text. If we want to link this primer and the widget a bit more, we may want to show how vaccination gets factored into this (how proportion susceptible enters the calculation). 
```suggestion
Then the NGM can be defined as $\mathbf{R}$ with elements $R_{ij} = \frac{\beta_{ij}N_i}{\gamma N}$. This is the formulation used for the input NGM in the widget, noting the implicit assumption that the user has provided entries to the input NGM that factor in population sizes. Vaccination alters the proportion of susceptible individuals that may become infected in each group, thus the rows of the input NGM are multiplied by the remaining proportion susceptible after vaccination. 
```

Author: Great suggestion, can add a section that speaks to the vaccination application too. I thought that the approach doc was sort of doing this already - what's something more fleshed out that you would want to see in the primer doc beyond what's in the approach doc?

Reviewer 3: You're right about the overlap here and I'm now recalling that this NGM primer was supposed to be more of a stand alone resource than the user guide and approach document and we had even discussed moving the NGM primer to the CFA recipes repo at one point? So maybe reject this suggestion in favor or keeping this primer very general?

Author: I think it's good to add this and help make this connection back to the model in the repo for anyone reading the primer.

# A Primer on Next Generation Matrix Models

> A Next Generation Matrix model is a way to model the expected number of infections generated by a typical infected individual in different groups or categories of the population in consecutive generations. The Next Generation Matrix (hereafter referred to as the NGM) encodes this information. NGM models are an effective way to model average dynamics in a heterogeneous population during the early growth phase and in the limit of the disease-free equilibrium.

Reviewer 3: ```suggestion
A Next Generation Matrix model is a way to model the expected number of infections that are generated by a typical infected individual in different groups or categories of the population in consecutive generations. The Next Generation Matrix (hereafter referred to as the NGM) encodes this information. NGM models are an effective way to model average dynamics in a heterogeneous population during the early growth phase and in the limit of the disease-free equilibrium.
```

As a result, most modelers familiar with NGMs have experience with using them as an analytical tool rather than as a simulation tool. However, NGMs can also be used to approximately model the ODEs for the subsystem of infected states.

## Interpretation of matrix elements
Imagine we have an NGM, $\mathbf{R} = \left(R_{ij}\right)$. The elements $R_{ij}$ of this matrix can be interpreted as the average number of infections in group $i$ caused by an infected individual in group $j$ between consecutive generations in a fully susceptible population. As a rule of thumb, the matrix $\mathbf{R}$ is not symmetric; some groups may be more susceptible to infection or more transmissive resulting in an asymmetric $\mathbf{R}$.

## Formal definition

> For a system of differential equations describing infectious disease dynamics, we can identify the infected subsystem that describes the production of new infections and other changes in state of infected individuals. After linearizing around the DFE, we can decompose the infected subsystem into 2 parts representating rates of transmission and transition. It is common to see the transmission component referred to as $\mathbf{T}$, the transmission matrix, and the transition component referred to as $\mathbf{\Sigma}$, the transition matrix. The Next Generation Matrix with Large domain is then defined as $\mathbf{R_L} = -\mathbf{T}\mathbf{\Sigma}^{-1}$.

Reviewer 3: maybe link to a definition of "large domain"?

* __Disease-free equilibrium__: The NGM is defined by identifying transmission and transition dynamics of an infectious disease model near the disease-free equilibrium (DFE) and linearizing the system around that point. A disease-free equilibrium is a point the epidemiological system where the population is free of disease, i.e., at a DFE the infectious population is zero. There can be multiple DFE for a system; the NGM is defined at the point where the population is fully susceptible. For example, in the classic SIR model, there exists a DFE with the conditions ($S \approx N$, $I \approx 0$, $R = 0$), which leads us to the condition $$R_0 = \frac{\beta}{\gamma} \geq 1$$ for growhth of disease in the population when we linearize the system around that point. Another DFE exists at the point where ($S = 0$, $I = 0$, $R = N$), however this DFE is not epidemiologically relevant to disease dynamics since disease cannot grow at this point.

> * __Depletion of susceptibles__: NGM models describe infectious disease dynamics as a demographic process in the sense that each consecutive generatino produces new offspring infections. This can be a good approximation for dynamics early on and in the limit of a large, otherwise fully susceptible population, such that stochastic effects are negligible. However, unlike ODE models, an NGM model does not account for the fixed size of a population and cannot model the depletion of susceptibles over time.

Reviewer 3: ```suggestion
* __Depletion of susceptibles__: NGM models describe infectious disease dynamics as a demographic process in the sense that each consecutive generation produces new offspring infections. This can be a good approximation for dynamics early on and in the limit of a large, otherwise fully susceptible population, such that stochastic effects are negligible. However, unlike ODE models, an NGM model does not account for the fixed size of a population and cannot model the depletion of susceptibles over time.
```
